utils/symbol.c

```diff
@@ -1430,3 +1423,71 @@ void print_symtabs(struct symtabs *symtabs)
 		symbol_putname(sym, name);
 	}
 }
+
+static unsigned long kaddr;
+
+static unsigned long get_kaddr(char *str)
+{
+	char *pch;
+	unsigned long addr;
+
+	pch = strtok (str,"-");
+	addr = strtoul(pch, NULL, 16);
+
+	if (addr < 0x40000000UL) {
+		return 0x40000000UL;
+	} else if (addr < 0x80000000UL) {
+		return 0x80000000UL;
+	} else if (addr < 0xB0000000UL) {
+		return 0xB0000000UL;
+	} else if (addr < 0xC0000000UL) {
+		return 0xC0000000UL;
+#if __SIZEOF_LONG__ == 8
+	} else {
+		return 0x800000000000ULL;
+	}
+#else
+    }
+#endif
+}
+
+void set_kaddr(char *dirname, const char *session_id)
+{
+	FILE *fp;
+	char buf[4096];
+	char line[200];
+
+	snprintf(buf, sizeof(buf), "%s/sid-%s.map", dirname, session_id);
+
+	fp = fopen(buf, "r");
+
+	if (fp == NULL)
+		exit(1);
+
+	while (fgets(line, sizeof(line), fp) != NULL) {
+		if (strstr(line, "[stack") != NULL) {
+			kaddr = get_kaddr(line);
+			break;
+		}
+	}
+	pclose(fp);
+}
+
+bool is_kernel_address(unsigned long addr)
+{
+	return addr >= kaddr;
+}
+
+unsigned long get_real_address(unsigned long addr)
+{
+	if (is_kernel_address(addr)) {
+#if __SIZEOF_LONG__ == 8
+		return (kaddr == 0x800000000000ULL) ? 
+                    (addr | (-1UL << KADDR_SHIFT)) : 
+                    (addr | kaddr);
+#else
+		return addr | kaddr;
+#endif
```

Comment:
This function is meaningless for 32-bit systems.  We save function addresses into 48-bit to save disk space.  So it should be extended to 64-bit on those systems to be check the address correctly.  However 32-bit systems don't have such issues so it's ok to return the given address as is.  The only thing needed to 32-bit systems is a correct `is_kernel_address()` function.